lib/cn/cn_tree.c

```diff
@@ -1211,6 +1218,31 @@ cn_tree_prepare_compaction(struct cn_compaction_work *w)
         goto err_exit;
     }
 
+    w->cw_vgmap = NULL;
+    if (kcompact || split) {
+        w->cw_vgmap = calloc(n_outs, sizeof(*w->cw_vgmap));
+        if (!w->cw_vgmap) {
+            err = merr(ENOMEM);
+            goto err_exit;
+        }
+
+        if (split) {
+            size_t sz = HSE_KVS_KEY_LEN_MAX +
+                n_outs * (sizeof(*(w->cw_split.commit)) + sizeof(*(w->cw_split.dgen))) +
+                w->cw_kvset_cnt * sizeof(*(w->cw_split.purge));
+
+            w->cw_split.key = calloc(1, sz);
+            if (!w->cw_split.key) {
+                err = merr(ENOMEM);
+                goto err_exit;
+            }
+
+            w->cw_split.commit = w->cw_split.key + HSE_KVS_KEY_LEN_MAX;
+            w->cw_split.purge = w->cw_split.commit + n_outs;
+            w->cw_split.dgen = (void *)(w->cw_split.purge + w->cw_kvset_cnt);
```

Comment:
I feel like the caller should have an array of pointers to set instead, otherwise the impl has to alloc/realloc a buffer of pointer to manage.

```
merr_t
shared_alloc(size_t ptrc, void ***ptrv, size_t sz, ...);
```

Then it doesn't need to be NULL terminated either.

Reply:
You can do two passes thorught the params with the form I posted, without any  extra allocs or reallocs, and the caller doesn't have to create a temp arrays to pass ptrs and sizes.

Reply:
Yea I guess I never thought of iterating through `va_list` twice.

Reply:
Should also include per pointer alignment information...

Reply:
Do we have any allocations like this where providing the alignment information would be useful? Can't think of any

Reply:
It makes sense to provide spacing between objects so that each one begins at an 8-byte boundary or something.  That should be automatic though.  If we want to support alignment, then it's just a matter of increasing the 8-byte boundary.

Reply:
Well I understand why do it, but I cannot recall any use cases where this API would be used that need memory alignment. I have only ever seen this pattern with malloc()